### PR TITLE
Add Store sharding and Receive hashrings

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

* Add `helpers.libsonnet`
* Update examples.

## Verification

* `make generate`

### CODE REVIEW

**Comment by metalmatze** [comment]
I don't see a need for this helpers file. The store sharding can be part of the store and similarly, the hashring can be part of the receviers.

**Review by metalmatze — CHANGES_REQUESTED** [review]

File: all.jsonnet [hunk 1]
```diff
@@ -149,10 +150,74 @@ local qf = t.queryFrontend(commonConfig {
...
+      // NOTICE: <MEMCACHED_SERCIVE> is a placeholder to generate examples.
+      // List of memcached addresses, that will get resolved with the DNS service discovery provider.
+      // For DNS service discovery reference https://thanos.io/service-discovery.md/#dns-service-discovery
+      addresses: ['dnssrv+_client._tcp.<MEMCACHED_SERCIVE>.%s.svc.cluster.local' % commonConfig.namespace],
+    },
+  },
+});
+
+local finalQ = q {
+  config+:: {
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply all.jsonnet:195]: This is nothing new. This was already there. I have actually changed my inital implementation to retrofit what we have for ruler.

**Comment by kakkoyun** [comment]
@metalmatze  Let's reach a consensus on this PR and then I'll open separate PRs for each different class of changes.

**Comment by kakkoyun** [comment]
To see it on the action: https://github.com/observatorium/deployments/pull/368

File: jsonnet/kube-thanos/kube-thanos-receive-hashrings.libsonnet [hunk 2]
```diff
@@ -0,0 +1,59 @@
+local receiveConfigDefaults = import 'kube-thanos/kube-thanos-receive-default-params.libsonnet';
+local receive = import 'kube-thanos/kube-thanos-receive.libsonnet';
+
+// These are the defaults for this components configuration.
+// When calling the function to generate the component's manifest,
+// you can pass an object structured like the default to overwrite default values.
+local defaults = {
```
Comment (metalmatze) [inline jsonnet/kube-thanos/kube-thanos-receive-hashrings.libsonnet:7]:
```suggestion
local defaults = receiveConfigDefaults {
```

No need to nest these defaults, I think :)

File: all.jsonnet [hunk 3]
```diff
@@ -149,10 +147,79 @@ local qf = t.queryFrontend(commonConfig {
   },
 });
 
+
+local hashrings = import 'kube-thanos/kube-thanos-receive-hashrings.libsonnet';
```
Comment (metalmatze) [inline all.jsonnet:151]: That should be necessary, when adding this to the thanos.libsonnet, like every other component. :)

File: all.jsonnet [hunk 4]
```diff
@@ -149,10 +147,79 @@ local qf = t.queryFrontend(commonConfig {
...
+  ],
+  receiveConfig: commonConfig {
+    replicas: 1,
+    replicationFactor: 1,
+    serviceMonitor: true,
+    hashringConfigMapName: 'hashring',
+  },
+});
+
+local shards = import 'kube-thanos/kube-thanos-store-shards.libsonnet';
```
Comment (metalmatze) [inline all.jsonnet:171]: Same. Add this to thanos.libsonnet and expose it as another component.

File: jsonnet/kube-thanos/kube-thanos-rule.libsonnet [hunk 5]
```diff
@@ -42,7 +42,7 @@ function(params) {
   // Combine the defaults and the passed params to make the component's config.
   config:: defaults + params,
   // Safety checks for combined config of defaults and params
-  assert std.isNumber(tr.config.replicas) && tr.config.replicas >= 0 : 'thanos receive replicas has to be number >= 0',
+  assert std.isNumber(tr.config.replicas) && tr.config.replicas >= 0 : 'thanos rule replicas has to be number >= 0',
```
Comment (metalmatze) [inline jsonnet/kube-thanos/kube-thanos-rule.libsonnet:45]: Ha! Oops. :sweat_smile:

**Review by metalmatze — CHANGES_REQUESTED** [review]

**Review by metalmatze — APPROVED** [review]
Let's merge it :)

**Pull request merged**